doc/code/event_system.md

```diff
@@ -0,0 +1,106 @@
+
+```
+t_old < t_e <= t_cur
+```
+
+where `t_cur` is the current simulation time, `t_old` is the last time `reach_time(..)` was
+called, and `t_e` is the event's scheduled invoke time. Events are executed in order of
+earliest to latest invoke time. During execution, events may trigger the rescheduling of
+other events if there are dependencies between the events.
+
+Both the creation and execution of events require passing a `State` object that functions
+as a persistant global storage for values that are not linked to a specific event executions.
+In practice, `State` is primarily used to pass the state of the [game simulation](/doc/code/game_simulation/)
+to the event system via an object of the derived `GameState` class. This object allows
+access to the indexing structures of the current game that can be used to retrieve
+the available game entities, for example.
+
+
+## Event Handler Setup
+
+The logic for (re-)scheduling and invoking an event is implemented by derivatives of the
+`EventHandler` class. Event handlers must implement 3 methods:
+
+* `invoke(..)`: Called by the event loop when the event is executed.
+* `predict_invoke_time(..)`: Calculates the desired invoke time. Called when the event is first created and whenever
+                           the event is due to be rescheduled (e.g. if a dependency event changes).
+* `setup_event(..)`: Sets up dependencies during creation of the event.
+
+Additionally, event handlers require a unique ID (defined as a string) as well as
+a *trigger type* which defines under which circumstances an event is executed.
+These parameters can be defined at runtime when instantiating the event handler.
+
+To enable events to use the event handler, an event handler object must first
+be initialized and then passed to the event loop. This can either be done explicitly via
+the `add_event_handler(..)` method of the `EventLoop` class or during the
+creation of an event via an overloaded `create_event(..)` method. Once an event
+handler is registered, new events can also be created using the event handler's
+unique ID string when calling `create_event(..)`.
+
+
+## Event Instances
+
+When calling `create_event(..)` on the event loop, an `Event` instance is returned which
+allows access to the configuration parameters of the event. Event instances store
+references to:
+
+* **Event payload**: Map of parameters passed to the event handler on invocation.
+* **Event target**: Entity targeted by the event. Implemented as a weak reference to let the target expire.
+* **Event handler**: Event handler object used for rescheduling and executing the event.
+* **Invoke time**: Determines when the event is executed.
+* **Time of last change**: Time of the last change to invocation time. Used for rescheduling the event.
+
+These parameters may be changed after creating the event. However, in most cases it's
+more beneficial to create a new event instead and [cancel](#canceling-events) the old
```

Comment:
in "most"? which are they? otherwise we could say "alternatively we can cancel and recreate"